lib/Kernel/RotateAndReduceFuzzTest.cpp

```diff
@@ -0,0 +1,175 @@
+  if (vec.empty()) return vec;
+  std::vector<int> result(vec.size(), 0);
+  for (int64_t i = 0; i < n; ++i) {
+    std::vector<int> rotated = rotate(vec, period * i);
+    for (size_t j = 0; j < vec.size(); ++j) {
+      result[j] += plaintexts[i][j] * rotated[j];
+    }
+  }
+  return result;
+}
+
+// Optimized implementation using the actual kernel
+std::vector<int> runImpl(const std::vector<int>& vec,
+                         const std::vector<std::vector<int>>& plaintexts,
+                         int64_t period, int64_t n,
+                         const std::string& reduceOp = "arith.addi") {
+  if (vec.empty()) return vec;
+  LiteralValue vectorInput(vec);
+
+  std::shared_ptr<ArithmeticDagNode<LiteralValue>> result;
+  std::optional<LiteralValue> plaintextsInput = std::nullopt;
+  if (!plaintexts.empty()) {
+    plaintextsInput = std::optional<LiteralValue>(LiteralValue(plaintexts));
+  }
+  result = implementRotateAndReduce(vectorInput, plaintextsInput, period, n,
+                                    reduceOp);
+  return std::get<std::vector<int>>(evalKernel(result).getTensor());
+}
+
+// Property: Implementation should match naive algorithm with plaintexts
+void RotateAndReduceWithPlaintextsMatchesNaive(
+    const std::vector<int>& vector, int64_t period, int64_t steps,
+    const std::vector<std::vector<int>>& plaintexts) {
+  // Ensure vector is non-empty
+  if (vector.empty() || steps <= 0 || period <= 0) return;
+
+  // Ensure plaintexts match the expected structure
+  if (plaintexts.size() != static_cast<size_t>(steps)) return;
+  for (const auto& plaintext : plaintexts) {
+    if (plaintext.size() != vector.size()) return;
+  }
+
+  std::vector<int> expected = runNaive(vector, plaintexts, period, steps);
+  std::vector<int> actual = runImpl(vector, plaintexts, period, steps);
+
+  EXPECT_EQ(expected, actual);
+}
+
+// Naive implementation for the no-plaintexts case
+std::vector<int> runNaiveNoPlaintexts(const std::vector<int>& vec,
+                                      int64_t period, int64_t steps) {
+  if (vec.empty()) return vec;
+  std::vector<int> result = vec;
+
+  // Sum up all the rotations
+  for (int64_t i = 1; i < steps; ++i) {
+    std::vector<int> rotated = rotate(vec, period * i);
+    for (size_t j = 0; j < vec.size(); ++j) {
+      result[j] += rotated[j];
+    }
+  }
+  return result;
+}
+
+// Property: Implementation without plaintexts should sum rotations
+void RotateAndReduceWithoutPlaintexts(const std::vector<int>& vector,
+                                      int64_t period, int64_t steps) {
+  // Ensure valid inputs
+  if (vector.empty() || steps <= 0 || period <= 0) return;
+
+  std::vector<int> expected = runNaiveNoPlaintexts(vector, period, steps);
+  std::vector<int> actual = runImpl(vector, {}, period, steps);
+
+  // Verify result matches the naive implementation
+  EXPECT_EQ(expected, actual);
+}
+
+// Property: Different reduce operations should produce different results
```

Comment:
The test doesn't end up asserting this property, but I wonder is there even any value in this test? Maybe a test with a muli reduce op should just be handled comparing it to a naive implementation